keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -613,7 +621,178 @@ sequence that it will not successfully decode.
+responsible for ensuring that that encoding of the slice contents matches the encoding that will be
+used to serialize the object containing the `RawExtension`. This is trivially ensured in cases 13
+and 14 because Protobuf is capable of representing `RawExtension` values containing arbitrary
+bytes. Protobuf is not a supported encoding for Unstructured objects. Existing programs have in
+practice stored JSON in the Raw field of `RawExtension`.
+
+#### Compatibility
+
+If the `RawExtension` marshalling and unmarshalling behavior for CBOR were to be implemented in
+exactly the same way as the existing JSON behaviors, the assumptions in many existing programs that
+the Raw field can be assigned to a slice of JSON bytes, or that the Raw bytes of an unmarshalled
+`RawExtension` are valid JSON, would be broken.
+
+The simple approach of automatically transcoding JSON to CBOR during CBOR marshalling, and
+transcoding CBOR to JSON during CBOR unmarshalling, would avoid breaking existing programs. However,
+the expense of transcoding to or from JSON would negate any performance advantage of a binary
+encoding. This expense would not be limited to a few API types: significant examples include the use
+of a `RawExtension` field in `metav1.WatchEvent` to represent each watch event's object state, or
+the arbitrary objects embedded in `admissionv1.AdmissionRequest`.
+
+Instead, automatic transcoding to and from JSON will become optional and initially enabled by
+default.
+
+For marshalling:
+
+1. If both Raw and Object are `nil`, `null` is returned.
+1. If Object is not `nil`, return the result of marshalling Object.
+1. If Raw is prefixed by the CBOR encoding of the head of the "self-described CBOR" tag, then return
+   the Raw bytes verbatim. This is the same prefix used by the CBOR `recognizer.RecognizingDecoder`
+   implementation, and all objects encoded by the CBOR `runtime.Encoder` implementation have this
+   prefix.
+1. Otherwise, assume Raw contains JSON and return the result of transcoding Raw from JSON to CBOR.
+
+Unmarshalling will behave the same for CBOR as it does for JSON and the input bytes will be copied
+verbatim to the Raw field. A Decoder implementation will be provided that delegates to another
+Decoder, then post-processes the result by transcoding `RawExtension` bytes to JSON before
+returning. The CBOR serializer available through `serializer.CodecFactory` will be wired to use
+this, allowing existing programs to continue to assume that unmarshalled Raw bytes contain JSON. The
+stream serializer will not. In practice, the watch decoder assumes that the non-stream serializer
+can directly decode the Raw bytes of a `metav1.WatchEvent` decoded by the stream serializer.
+
+##### Migration
+
+**GA**
+
+*Naive Clients*
+
+1. Client assumes received RawExtension is JSON.
+1. Client receives CBOR response body. The response bytes that represent the RawExtension are CBOR.
```

Comment:
I really like this "1:1 match the body content type with the RawExtension content type".  Should we call this principal out more loudly anywhere in this section?

Reply:
Good idea. As I understand it, this rule is implied for any protocol that supports Unstructured. It's just been hidden because Unstructured has always meant JSON.

Reply:
for types that are built in on the server side that have RawExtension, the server cannot always convert RawExtension data to match the body content type

for example, the server cannot convert ControllerRevision.Data to proto when returning a proto response

Reply:
For the proto case did we agree to always use JSON?  I vaguely recall this being discussed in a sig meeting.

Reply:
So the rule mighttt be:  For JSON and CBOR, we always encode RawExtension to match the content type of the surrounding body, for Protobuf, we always encode JSON?

Reply:
I think the minimum requirement for RawExtension inside Protobuf has to be that the content can be unmarshallable to Unstructured. Otherwise, there could be objects that can only be representable in Protobuf.

I don't know if that opinion is shared by anyone else. It might be worth a (shorter) discussion at a biweekly meeting now that this KEP has raised the topic.